### PR TITLE
DAT-3362 Identify disambiguation and context-link templates

### PR DESCRIPTION
**Pull request by rybmat**
https://wikia-inc.atlassian.net/browse/DAT-3363

Identify disambiguation and context-link templates

### CODE REVIEW

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+  private Boolean matchTitle(Document document) {
+    String title = document.title().toLowerCase();
+    return title.contains("youmay")
+           || title.contains("otheruse")
+           || title.contains("disambig")
+           || title.contains("неоднозначность")
+           || title.equals("template:for")
+           || title.equals("template:ambig")
```
Comment (WikiaTech) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java]: MAJOR Move the "template:ambig" string literal on the left side of this string comparison. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1132)

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+           || title.equals("template:dis")
```
Comment (WikiaTech) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java]: MAJOR Move the "template:dis" string literal on the left side of this string comparison. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1132)

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+           || title.equals("template:dislk");
```
Comment (WikiaTech) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java]: MAJOR Move the "template:dislk" string literal on the left side of this string comparison. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1132)

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+  private Boolean matchClass(Document document) {
+    return !document.select("*[class*=disambig], "
+                            + "*[class*=dablink]").isEmpty();
+  }
+
+  private Boolean matchTitle(Document document) {
+    String title = document.title().toLowerCase();
+    return title.contains("youmay")
```
Comment (WikiaTech) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java]: MAJOR Reduce the number of conditional operators (7) used in the expression (maximum allowed 3). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+           || title.contains("otheruse")
+           || title.contains("disambig")
+           || title.contains("неоднозначность")
+           || title.equals("template:for")
```
Comment (WikiaTech) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java]: MAJOR Move the "template:for" string literal on the left side of this string comparison. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1132)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/guice/TemplateClassificationWorkerModule.java [hunk 6]
```diff
@@ -26,6 +27,7 @@ protected void configure() {
 
     matchers.addBinding().to(NavboxMatcher.class);
     matchers.addBinding().to(QuoteMatcher.class);
+    matchers.addBinding().to(DisambiguationMatcher.class);
     matchers.addBinding().to(ScrollboxMatcher.class);
     matchers.addBinding().to(PortableInfoboxMatcher.class);
     matchers.addBinding().to(CustomInfoboxMatcher.class);
```
Comment (SebastianMarzjan) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/guice/TemplateClassificationWorkerModule.java]: Let's sort those too :)

File: workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java [hunk 7]
```diff
@@ -27,6 +27,7 @@
   public static final String NAVBOX_TYPE = "navbox";
   public static final String QUOTE_TYPE = "quote";
   public static final String SCROLLBOX_TYPE = "scrollbox";
+  public static final String DISAMBIGUATION_TYPE = "disambiguation";
```
Comment (SebastianMarzjan) [inline workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java]: I think it's time to sort those too by the const name

File: workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java [hunk 8]
```diff
@@ -48,6 +49,13 @@ public void testTemplateOther() {
   }
 
   @Test
+  public void testTemplateDisambiguation() {
```
Comment (SebastianMarzjan) [inline workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java]: Please consider tests for other cases too

Comment (rybmat) [inline workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java]: just test every template name that is checked in code?

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java (hunk 4 above)
Comment (WikiaTech) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java]: MAJOR Reduce the number of conditional operators (7) used in the expression (maximum allowed 3). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Either log or rethrow this exception.](https://github.com/Wikia/pandora/blob/f18d45700cb514e8d0ec347782291e42231fd13b/workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java#L209) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java (hunk 8 above)
Comment (rybmat) [inline workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java]: done

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    return this.matchClass(document)
+           || this.matchTitle(document);
+  }
+
+  private Boolean matchClass(Document document) {
+    return !document.select("*[class*=disambig], "
+                            + "*[class*=dablink]").isEmpty();
+  }
+
+  private Boolean matchTitle(Document document) {
```
Comment (WikiaTech) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java]: MAJOR The Cyclomatic Complexity of this method "matchTitle" is 14 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Either log or rethrow this exception.](https://github.com/Wikia/pandora/blob/ab6f9dee58e9ceb020cde8b2cdfc948c80ed9806/workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java#L209) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+           || title.contains("неоднозначность")
+           || title.contains("distinguish")
+           || title.contains("redirect")
+           || title.contains("looking for")
+           || title.contains("rhere")
+           || "template:confuse".equals(title)
+           || "template:for".equals(title)
+           || "template:ambig".equals(title)
+           || "template:dis".equals(title)
+           || "template:also".equals(title)
```
Comment (dianafa) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java]: The `also` template shouldn't be a Context Link ?

Comment (SebastianMarzjan) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/DisambiguationMatcher.java]: May be. Let's merge the code for both filters into a single branch and compare.

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/ContextLinkMatcher.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  /**
+   * check if title of document is characteristic to a typical Context Link name.
+   *
+   * @param document template's HTML
+   * @return boolean
+   */
+  private Boolean matchTitle(Document document) {
+    String title = document.title().toLowerCase();
```
Comment (WikiaTech) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/ContextLinkMatcher.java]: MAJOR Remove this unused "title" local variable. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1481)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Either log or rethrow this exception.](https://github.com/Wikia/pandora/blob/428c6f0655cd92815b5daeb5d93730a1f8c9eed1/workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java#L233) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java [hunk 12]
```diff
@@ -171,6 +172,71 @@ public void testTemplateWithWikitextTable() {
...
+
+  @Test
+  public void testTemplateContextLinkByClass() {
+    Set<String> expected = new HashSet<>(Collections.singletonList(CONTEXT_LINK_TYPE));
+    assertEquals(expected,
+                 this.test(this.getStringFromResource("/html/TemplateContextLinkByClass.html")));
+  }
+
+  @Test
+  public void testTemplateContextLinkSee_also() {
```
Comment (WikiaTech) [inline workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java]: MAJOR Rename this method name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00100)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Either log or rethrow this exception.](https://github.com/Wikia/pandora/blob/c2c3f7e29078d5f5dde5ddac3d8a9ea25a0244ad/workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java#L246) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/guice/TemplateClassificationWorkerModule.java [hunk 13]
```diff
@@ -24,11 +25,15 @@ protected void configure() {
...
+    matchers.addBinding().to(CustomInfoboxMatcher.class);
+    matchers.addBinding().to(CustomReferences.class);
     matchers.addBinding().to(NavboxMatcher.class);
+    matchers.addBinding().to(PortableInfoboxMatcher.class);
     matchers.addBinding().to(QuoteMatcher.class);
     matchers.addBinding().to(ScrollboxMatcher.class);
     matchers.addBinding().to(PortableInfoboxMatcher.class);
     matchers.addBinding().to(CustomInfoboxMatcher.class);
     matchers.addBinding().to(CustomReferences.class);
+    matchers.addBinding().to(ContextLinkMatcher.class);
```
Comment (idradm) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/guice/TemplateClassificationWorkerModule.java]: There is some chaos going on here, same bindings, not sorted, could you take a look and fix it up?

Comment (dianafa) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/guice/TemplateClassificationWorkerModule.java]: true

File: workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/ContextLinkMatcher.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
+package com.wikia.workers.templateclassificationworker.matchers;
+
+import org.jsoup.nodes.Document;
+
+public class ContextLinkMatcher implements TemplateTypeMatcher {
```
Comment (idradm) [inline workers/template-classification-worker/src/main/java/com/wikia/workers/templateclassificationworker/matchers/ContextLinkMatcher.java]: Java docs don't add any new information here, especially when they are used for privates methods. Do we need them?

File: workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java [hunk 15]
```diff
@@ -21,12 +21,13 @@
...
   public static final String CUSTOM_INFOBOX_TYPE = "custom-infobox";
-  public static final String OTHER_TYPE = "other";
   public static final String INFOBOX_TYPE = "infobox";
-  public static final String REFERENCES_TYPE = "references";
   public static final String NAVBOX_TYPE = "navbox";
+  public static final String OTHER_TYPE = "other";
   public static final String QUOTE_TYPE = "quote";
+  public static final String REFERENCES_TYPE = "references";
   public static final String SCROLLBOX_TYPE = "scrollbox";
+  public static final String CONTEXT_LINK_TYPE = "context-link";
```
Comment (idradm) [inline workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java]: Could we sort them for better readability?

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Either log or rethrow this exception.](https://github.com/Wikia/pandora/blob/9dbc72ee91e71af67acf3cfc4c353e6b09bf2fec/workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java#L253) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

**Comment by idradm** [comment]
LGTM

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Either log or rethrow this exception.](https://github.com/Wikia/pandora/blob/aa1a51dc8677d3b6b542104a0eda01c2f8e50924/workers/template-classification-worker/src/test/java/com/wikia/workers/templateclassificationworker/TemplateParserTest.java#L253) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

**Comment by dianafa** [comment]
LGTM

**Pull request merged**